nu_stats/fqutils.py

```diff
@@ -0,0 +1,772 @@
+        idx = np.linspace(self._min_index,
+                 self._max_index,10000)
+        ll = [self(E, index, interp) for index in idx]
+        if interp and not scaled_plot:
+            ll = [l*self._Nbins for l in ll]
+        p = plt.plot(idx, ll, **kwargs)
+        return p
+
+    def plot_pdf_meshgrid(self, interp=2, **kwargs):
+        idx = np.linspace(self._min_index,
+                 self._max_index,100)
+        Es = np.logspace(np.log10(self._min_E),
+                 np.log10(self._max_E),100)
+        Es, idx = np.meshgrid(Es, idx)
+        vfunc = np.vectorize(self.__call__)
+        Z = vfunc(Es, idx, interp)
+        # np.savetxt("pdfmesh.csv", Z, delimiter=",")
+        plt.pcolormesh(Es, idx, Z, shading='auto', **kwargs)
+        plt.xscale('log')
+
+
+class FqStructure:
+    def __init__(self):
+        """
+        Class for holding useful stuff to do with the Braun paper approach.
+        """
+        self.spacial_only = True # Spacial only until energy likelihoods are given
+    
+    def construct_source_energy_likelihood(self,
+        z: float,
+        Emin: u.GeV,
+        Emax: u.GeV,
+        min_index: float,
+        max_index: float,
+        n_Esim: int,
+        Nbins: int,
+        prefab_likelihood_file: str = None
+    ):
+        print('Generating marginalized energy likelihood..')
+        '''
+        Because we only want a sample of Edet for a certain gamma, with no
+        regard for spacial factor, we can use a single source with any L.
+        '''
+        self.energy_likelihood = MarginalisedEnergyLikelihood(
+            is_source = True,
+            z = z,
+            Emin = Emin.value,
+            Emax = Emax.value,
+            n_Esim = n_Esim,
+            min_index = min_index,
+            max_index = max_index,
+            Nbins = Nbins,
+            prefab_file = prefab_likelihood_file
+        )
+        self.spacial_only = False
+        print('Marginalized energy likelihood generated.')
+
+    def construct_diffuse_bg_energy_likelihood(self,
+        z: float,
+        Emin: u.GeV,
+        Emax: u.GeV,
+        min_index: float,
+        max_index: float,
+        n_Esim: int,
+        Nbins: int,
+        prefab_likelihood_file: str = None
+    ):
+        print('Generating marginalized background energy likelihood..')
+        '''
+        TODO: Because is_source=False here will make the reference
+        energy simulations be set to bg only and z in the Simulation input is
+        the source's redshift and the bg is hardcoded, z here won't actually do
+        anything, but can be good to keep in case of restructuring.
+        This comment is a warning that passing another z to the bg_EL will not
+        necessarily do anything.
+        As of 27th of March 1.0 is hardcoded as z_bg in Simulation so that is
+        the redshift that will go into the bg_MEL.
+        '''
+        self.bg_energy_likelihood = MarginalisedEnergyLikelihood(
```

Comment:
Agree with your TODO, this is probably worth sorting out soonish so that you don't have to worry about it when experimenting.